File: Java/cp/Roman_To_Integer.java
// Program to convert Roman Numeral to Integer


import java.util.Scanner;

class Solution {
              
    public static int symboltoChar(char c){
        switch(c){
            case 'I': return 1;
            case 'V' : return 5;
            case 'X' : return 10;
            case 'L': return 50;
            case 'C' : return 100;
            case 'D' : return 500;
            case 'M' : return 1000;
        }
        return 0;
    }

    public static void main(String...ar){
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter Roman Numeral");
        String s= sc.next();
        int result=0;
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            
            if((i!= s.length()-1) && (symboltoChar(c) < symboltoChar(s.charAt(i+1)))){
                result= result - symboltoChar(c);
            }else{
                result = result + symboltoChar(c);
            }
               }
               System.out.println(result);
        
    }
}

File: C-Plus-Plus/math/Roman_To_Integer.cpp
/*
Introduction 
Given a roman numeral, convert it to an integer.

Argument/Return Type
Given a string which represents a roman numeral ,
return its corresponding decimal integer value
*/

#include <bits/stdc++.h> 
using namespace std;

//Function to convert the given roman numeral to its corresponding integer value
int romanToInt(string roman) 
{ 
    //Define a map     
    map<string,int>value;
    
    //Store the integer values of all valid roman symbols
    value["I"]  = 1;
    value["IV"] = 4;
    value["V"]  = 5;
    value["IX"] = 9;
    value["X"]  = 10;
    value["XL"] = 40;
    value["L"]  = 50;
    value["XC"] = 90;
    value["C"]  = 100;
    value["CD"] = 400;
    value["D"]  = 500;
    value["CM"] = 900;
    value["M"]  = 1000;
    
    //Now traverse over all characters of string
    int index=0;
    //Initiate the decimal value with 0
    int decimalValue=0;

    while(index<roman.length())
    {
        //If there is one more char after the current position
        if(index<roman.length()-1)
        {
            /* check if the substring corresponding to two chars 
            is a valid roman symbol */
            string sub=roman.substr(index,2);
            if(value[sub])
            {
                /* If substring is a valid roman symbol 
                 add it to decimal value and move the index forward two times */
                decimalValue=decimalValue+value[sub];
                index=index+2;
            }
            else
            {
                // else add the decimal value the char at this position 
                sub=roman.substr(index,1);
                decimalValue=decimalValue+value[sub];
                index=index+1;
            }
        }
        else
        {
            /* If current position is end of the string
             add the decimala value of this char and continue */
            string sub=roman.substr(index,1);
            decimalValue=decimalValue+value[sub];
            index=index+1;
        }
        
    }
    
    return decimalValue;
}


// Driver code
int main()
{
    //Take input of a roman numeral string
    string roman;
    cout<<"Enter a roman numeral : ";
    cin>>roman;

    //Call the function and print the result
    cout<<"Hence the integer value of given roman numeral is ";
    cout<<romanToInt(roman);

    return 0; 
}
    
/*
Input:
1 <= roman.length <= 15
roman contains only the characters ('I', 'V', 'X', 'L', 'C', 'D', 'M').
It is guaranteed that s is a valid roman numeral in the range [1, 3999].

Sample Test Case 1  

Input Format : 
Example :
Enter a roman numeral : MCMXCIV

Output Format :
Example : ( Output to the above input example ) 
Hence the integer value of given roman numeral is 1994

Time/Space Complexity
Time Complexity : O(n) 
Space Complexity : O(1) 
*/
